# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



Оценка влияния различных факторов на платежеспособность заемщика.

Цель проекта: 

определить влияение на факт погашения кредита в срок следующих факторов:
 - семейное положение,
 - и количество детей клиента,
 - уровень дохода,
 - цель кредита.

Входные данные от банка — статистика о платёжеспособности клиентов.


Описание данных

children — количество детей в семье <br>
days_employed — общий трудовой стаж в днях <br>
dob_years — возраст клиента в годах <br>
education — уровень образования клиента <br>
education_id — идентификатор уровня образования <br>
family_status — семейное положение <br>
family_status_id — идентификатор семейного положения<br>
gender — пол клиента<br>
income_type — тип занятости<br>
debt — имел ли задолженность по возврату кредитов<br>
total_income — ежемесячный доход<br>
purpose — цель получения кредита

## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd


In [2]:
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [23]:
#сгруппируем заемщиков по количеству детей
report_one = data.groupby('children').agg({'debt': ['count','sum']})
report_one['debt_ratio'] = report_one['debt']['sum']/report_one['debt']['count'] ##доля заемщиков, которые имели задолженность, непогашенную в срок
report_one['client_ratio'] = report_one['debt']['sum']/report_one['debt']['sum'].sum() #доля каждой группы заемщицов в общем количестве заемщиков
report_one


debt       debt_ratio client_ratio
          count   sum                        
children                                     
0         14091  1063   0.075438     0.613741
1          4808   444   0.092346     0.256351
2          2052   194   0.094542     0.112009
3           330    27   0.081818     0.015589
4            41     4   0.097561     0.002309
5             9     0   0.000000     0.000000

**Вывод:** 
Заемщики с детьми чаще возвращают кредит позже срока. 
При этом заемщики с тремя детьми возвращают кредит позже срока реже, чем заемщики с одним ребенком или двумя детьми.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
#сгруппируем заемщиков по семейному положению
report_two = data.groupby('family_status').agg({'debt': ['count','sum']})
report_two['debt_ratio'] = report_two['debt']['sum']/report_two['debt']['count'] #доля заемщиков, которые имели задолженность, непогашенную в срок
report_two['client_ratio'] = report_two['debt']['sum']/report_two['debt']['sum'].sum() #доля каждой группы заемщиков в общем количестве заемщиков
report_two.sort_values('debt_ratio')

debt      debt_ratio client_ratio
                       count  sum                        
family_status                                            
вдовец / вдова           951   63   0.066246     0.036374
в разводе               1189   84   0.070648     0.048499
женат / замужем        12261  927   0.075606     0.535219
гражданский брак        4134  385   0.093130     0.222286
Не женат / не замужем   2796  273   0.097639     0.157621

In [25]:
#вариант решения с использованием функции
def func(factor):
    report = data.groupby(factor).agg({'debt': ['count','sum']})
    report['debt_ratio'] = report['debt']['sum']/report['debt']['count'] 
    report['client_ratio'] = report['debt']['sum']/report_two['debt']['sum'].sum()
    return report

func(factor=['family_status'])

debt      debt_ratio client_ratio
                       count  sum                        
family_status                                            
Не женат / не замужем   2796  273   0.097639     0.157621
в разводе               1189   84   0.070648     0.048499
вдовец / вдова           951   63   0.066246     0.036374
гражданский брак        4134  385   0.093130     0.222286
женат / замужем        12261  927   0.075606     0.535219

**Вывод:** 
Доля заемщиков, имеющих случаи непогашения кредита в срок, выше среди не состоящих в браке(в том числе и состоящих в гражданском браке), чем у заемщиков, состоящих в браке.Наибольшая доля заемщиков, имеющих случаи несвоервеменных погашений кредита, приходится на не состоящих в официальном браке заемщиков, и составляет 9,76% и 9,53% для не женатых/не замужних и состаящих в гражданском браке заемщиков соответсвенно, что на 2,2 п.п. и 1,97 п.п больше, чем аналогичный показатель среди состоящих в браке заемщиков (7,56%) соответсвенно. 

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [26]:
#сгруппируем заемщиков по уровню дохода
report_three = data.groupby('total_income_category').agg({'debt': ['count','sum']})
report_three['debt_ratio'] = report_three['debt']['sum']/report_three['debt']['count'] #доля заемщиков, которые имели задолженность, непогашенную в срок
report_three['client_ratio'] = report_three['debt']['sum']/report_three['debt']['sum'].sum() #доля каждой группы заемщиков в общем количестве заемщиков
report_three.sort_values('total_income_category')

debt       debt_ratio client_ratio
                       count   sum                        
total_income_category                                     
A                         25     2   0.080000     0.001155
B                       5014   354   0.070602     0.204388
C                      15921  1353   0.084982     0.781178
D                        349    21   0.060172     0.012125
E                         22     2   0.090909     0.001155

**Вывод:** 
98,56% рассматриваемых заемщиков относятся к двум категорям:
- с доходом от 50 000 до 200 000 руб. (78,12%, категория С)
- с доходом от 200 000 до 1 000 000 руб. (20,44%, категория B).
Остальные категории составляют 1,2 % и менее и являются статистически незначимымми.

Заемщики с более низкими доходами (50 000 - 200 000 руб.) на 14,3% чаще погашали задолженность позже срока.



#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [27]:
#сгруппируем заемщиков по цели кредита
report_four = data.groupby('purpose_category').agg({'debt': ['count','sum']})
report_four['debt_ratio'] = report_four['debt']['sum']/report_four['debt']['count'] #доля заемщиков, которые имели задолженность, непогашенную в срок
report_four['client_ratio'] = report_four['debt']['sum']/report_four['debt']['sum'].sum() #доля каждой группы заемщиков в общем количестве заемщиков
report_four

debt      debt_ratio client_ratio
                          count  sum                        
purpose_category                                            
операции с автомобилем     4279  400   0.093480     0.230947
операции с недвижимостью  10751  780   0.072551     0.450346
получение образования      3988  369   0.092528     0.213048
проведение свадьбы         2313  183   0.079118     0.105658

In [28]:
#сформируем сводную таблицу по целям кредитования и семейному статусу и найдем долю несвоевременных платежей по каждой группе
pivot_report_debt = data.pivot_table(index=['purpose_category'], columns=['family_status'],values='debt',aggfunc='sum')
pivot_report_count=  data.pivot_table(index=['purpose_category'], columns=['family_status'],values='debt',aggfunc='count')
pivot_ratio = pivot_report_debt/pivot_report_count
pivot_ratio

family_status             Не женат / не замужем  в разводе  вдовец / вдова  \
purpose_category                                                             
операции с автомобилем                 0.129134   0.075269        0.091743   
операции с недвижимостью               0.081337   0.068452        0.052336   
получение образования                  0.107826   0.071429        0.075758   
проведение свадьбы                          NaN        NaN             NaN   

family_status             гражданский брак  женат / замужем  
purpose_category                                             
операции с автомобилем            0.118605         0.083180  
операции с недвижимостью          0.092105         0.069727  
получение образования             0.148883         0.083528  
проведение свадьбы                0.079118              NaN

**Вывод:** 
Наименьшая доля заемщиков, имеющих случаи несвоевременного погашения кредита, относится к кредитам на операции с недвижимостью (7.26%). Следует отметить, что возврат в срок кредита на операции с недвижимостью имеет наименьшую долю просрочек среди всех категорий заемщиков, независимо от семейного положения. Исключение составляют заемщики в гражданском браке, которые добросовестней всего отдают кредит на проведение свадьбы (7,91%).
Кредиты на операции с атомобилем и получение образования имеют приблизительно одинаковое количество несвоевременных погашений - 9,35% и 9,25% соответственно. 




#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

**Ответ:** 
Пропуски в исходных данных имели место в двух столбцах, отражающих ежемесячный стаж и среднемесячный доход. При этом, там где был не указан доход, там же был не указан и стаж. Возможными причинами могла быть техническая ошибка при выгрузке данных. Либо по данным заемщикам не были изначально указаны данные о доходе и стаже. 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

**Ответ:** 
Медианное значение невилирует выбросы в данных, наличие которых может значительно отразиться на подсчете, например, среднего арифмитического значения. 

### Шаг 4: общий вывод.

**Вывод**
Для исследования надежности заемщиков были проанализированы предоставленные данные. 
Предварительный анализ данных показал наличие пропусков в данных по 2174 заемщикам, отражающих ежемесячный стаж и среднемесячный доход. Возможными причинами могла быть техническая ошибка при выгрузке данных. Либо по данным заемщикам не были изначально указаны данные о доходе и стаже. Для проведения анализа пропуски были заменены на медианные значения, соответствующие категориям заемщиков по занимаемой должности.
Кроме того были обнаружены 71 дубликат. Данные строки были удалены.

В результате проведенной раюоты сделаны следующие выводы:
1. Заемщики с детьми чаще возвращают кредит позже срока. При этом заемщики с тремя детьми возвращают кредит позже срока реже, чем заемщики с одним ребенком или двумя детьми.
2. Наибольшая доля заемщиков, имеющих случаи несвоервеменных погашений кредита, приходится на не состоящих в официальном браке заемщиков, и составляет 9,76% и 9,53% для не женатых/не замужних и состаящих в гражданском браке заемщиков соответсвенно, что на 2,20 п.п. и 1,97 п.п больше, чем аналогичный показатель среди состоящих в браке заемщиков (7,56%) соответсвенно. 

3. 98,56% рассматриваемых заемщиков относятся к двум категорям:
- с доходом от 50 000 до 200 000 руб. (78,12%, категория С)
- с доходом от 200 000 до 1 000 000 руб. (20,44%, категория B).
Заемщики с более низкими доходами (50 000 - 200 000 руб.) на 14,3% чаще погашали задолженность позже срока.

4. Наименьшая доля заемщиков, имеющих случаи несвоевременного погашения кредита, относится к кредитам на операции с недвижимостью (7.26%). Следует отметить, что возврат в срок кредита на операции с недвижимостью имеет наименьшую долю просрочек среди всех категорий заемщиков, независимо от семейного положения. Исключение составляют заемщики в гражданском браке, которые добросовестней всего отдают кредит на проведение свадьбы (7,91%).
Кредиты на операции с автомобилем и получение образования имеют приблизительно одинаковое количество несвоевременных погашений - 9,35% и 9,25% соответственно.

Таким образом, можно сделать вывод, что на погашение кредита в срок влияют разные факторы, в том числе семейное положение, наличие и количество детей, доход заемщиков, а так же цели кредитования. Для оценки надежности потенциального заемщика все эти факторы необходимо рассматривать в совокупности.